### ISSUE TITLE
Ambiguous definitions (in several headers) lead to "use of undeclared identifier" error

### ISSUE THREAD

**Issue by TediumRemedy** [niosus/EasyClangComplete#333]
This is not an error or bug report, just a question.
I am trying to set up an environment for micro controller development (STM32, using STM32CubeMX-generated Makefile-based project for a GCC compiler (and I put the arm-none-eabi-gcc compiler itself in the project directory). This project contains all the necessary header/source files in different subdirectories, it is self-contained and does not use any external libraries/headers).
I have created a file .clang_complete, specifying paths to all header subdirectories within my project dir.
For some reason, EasyClangComplete shows errors for definitions that are declared in multiple files, as on this picture:
![screen shot 2017-10-06 at 5 01 36 am](https://user-images.githubusercontent.com/5770542/31259970-7d09aca4-aa53-11e7-9a72-85080a89b3bb.png)
 
When I put mouse cursor on this GPIOA definition, I see that it is found in multiple header files (of course, in the project itself only one of them is used, they are just for different MCU families):
![screen shot 2017-10-06 at 5 02 07 am](https://user-images.githubusercontent.com/5770542/31260016-aa718824-aa53-11e7-883b-b409abd2b876.png)
And eventually this:
![screen shot 2017-10-06 at 5 03 30 am](https://user-images.githubusercontent.com/5770542/31260025-bcf0f200-aa53-11e7-9882-1bdaeb61494c.png)

Is there anything I can do to resolve this issue? I am using GCC, if that matters.
Thank you

Sublime: 3.0
System: macOS

**Comment by ubidefeo** [comment]
I kinda bumped into the same issue while trying to build particle.io (also based on STM32) projects in ST3.
I had to go manually scavenge every `inc` folder and place it in `.clang_complete` but after that the errors related to multiple declarations went away.
I also did a few "Clean current CMake cache".

As I kept adding/cleaning, the alerts disappeared. 
I noticed that while I had several declarations in different headers, the "confusion" only seemed to be there until you found the right declaration file. I think what you see is kind of guessing.
In particular for me it happened with the `Serial` object in the Arduino framework and its declarations of `print` and `println`. 
It only disappeared once I added the right headers for `Print.h` (which was being used by Particle's USB Serial abstraction.

Hope it helps

**Comment by TediumRemedy** [comment]
@ubidefeo Thanks for the response. So basically you added each and every .h file into .clang_complete except for those that may redefine already defined methods/macros?

**Comment by ubidefeo** [comment]
@TediumRemedy 
I didn't. I added the top-level include folders.
It might be helpful for you to check out [Particle.io's repository](https://github.com/spark/firmware)

the Makefile refers to a particular folder structure so in my blink test `main.cpp` file I needed some headers to be found for methods to resolve (`Serial.println` in the specific case).

I added the `inc` folders one by one (not the single `.h` files) and after every new entry I issues a `Clean CMake cache`.

Probably the following will be of no help to you, but you never know :)

this is what my `.clang_complete` file looks like
```
-I../firmware/wiring/inc
-I../firmware/hal/src/electron/modem
-I../firmware/hal/shared
-I../firmware/services/inc
-I../firmware/system/inc
-I../firmware/user/inc
-I../firmware/libraries/inc
-I../firmware/hal/inc
-I../firmware/system/inc
```

The folder with my `main.cpp` is at the same level of Particle's firmware folder.
The `make` command is issued from within folder `firmware\modules` as described in [this guide](https://docs.particle.io/faq/particle-tools/local-build/electron/#building).

**Comment by niosus** [comment]
Hey guys, sorry for not responding to this earlier. This is clearly just a configuration issue, so I will close it here, but we can continue discussing the possible ways to solve this here.

As @ubidefeo has pointed out, the errors will go away if you provide the correct folders that contain the headers. For example, let's say there is a file `/home/blah/include/cool/file.h` in your system and in your code there is a line ```c++
#include "cool/file.h"
```
To make sure clang can find the header you need to provide the flag `"-I/home/blah"` to tell clang where to search for headers. If your code does not use cmake not does it use a compilation database, there is nothing better that you can do except of manually specifying these. 

Alternatively, you can try Bear to generate a compilation database from the make file. You can then directly use this database with the plugin without manually specifying the flags.

**Comment by TediumRemedy** [comment]
@niosus 
Igor, could you please check the other issue, that describes the problem: https://github.com/niosus/EasyClangComplete/issues/424
Please note that I have instructed easyclangcomplete with all the directories that contain header files in my project using a .clang_complete file.
I have also created a demo repository with my project for you to check the issue: https://github.com/TediumRemedy/SonarTest

In short, the problem is that macros, that are defined in more than 1 header file, are considered undefined by the EasyClangComplete. 

Thanks for you help!
